File: apps/api/engine/parser.py
import json
import re
import hashlib
from pydantic import BaseModel
from typing import Optional, Union

AutoApproveValue = Union[bool, str, list[str]]


class MCPServer(BaseModel):
    name: str
    command: Optional[str] = None
    args: list[str] = []
    env: dict[str, str] = {}
    headers: dict[str, str] = {}
    url: Optional[str] = None
    transport: Optional[str] = None
    auto_approve: Optional[AutoApproveValue] = None
    disabled: bool = False


class MCPConfig(BaseModel):
    config_hash: str
    servers: list[MCPServer]


def _strip_jsonc_comments(text: str) -> str:
    """Strip // line comments and /* */ block comments from JSONC text.

    Fully string-aware state machine: escape sequences inside strings are
    skipped as a pair so `"\\"` (escaped backslash) never incorrectly
    toggles the in-string state.
    """
    result: list[str] = []
    in_string = False
    i = 0
    while i < len(text):
        ch = text[i]
        if in_string:
            if ch == '\\' and i + 1 < len(text):
                # Escaped character — consume both chars without toggling state
                result.append(ch)
                result.append(text[i + 1])
                i += 2
                continue
            if ch == '"':
                in_string = False
            result.append(ch)
        else:
            if ch == '"':
                in_string = True
                result.append(ch)
            elif ch == '/' and i + 1 < len(text):
                if text[i + 1] == '/':
                    # Line comment — skip to end of line
                    while i < len(text) and text[i] != '\n':
                        i += 1
                    continue
                elif text[i + 1] == '*':
                    # Block comment — skip to */
                    i += 2
                    while i < len(text) - 1:
                        if text[i] == '*' and text[i + 1] == '/':
                            i += 2
                            break
                        i += 1
                    continue
                else:
                    result.append(ch)
            else:
                result.append(ch)
        i += 1
    return ''.join(result)


def _extract_first_json_object(text: str) -> str:
    """Return the first complete {...} block from text (handles multi-object JSONC files).

    String-aware: does not count { or } that appear inside quoted strings.
    """
    depth = 0
    in_string = False
    start = -1
    i = 0
    while i < len(text):
        ch = text[i]
        if ch == '\\' and in_string:
            i += 2  # skip escaped character entirely
            continue
        if ch == '"':
            in_string = not in_string
        elif not in_string:
            if ch == '{':
                if depth == 0:
                    start = i
                depth += 1
            elif ch == '}':
                depth -= 1
                if depth == 0 and start != -1:
                    return text[start:i + 1]
        i += 1
    return text[start:] if start != -1 else text


def _parse_headers(raw: object) -> dict[str, str]:
    if not isinstance(raw, dict):
        return {}
    return {str(k): str(v) for k, v in raw.items()}


def _parse_auto_approve(server_data: dict) -> Optional[AutoApproveValue]:
    """Parse autoApprove / alwaysAllow from Cursor or Claude Desktop configs."""
    raw = server_data.get("autoApprove")
    if raw is None:
        raw = server_data.get("alwaysAllow")
    if raw is None:
        return None
    if isinstance(raw, bool):
        return raw
    if isinstance(raw, str):
        return raw
    if isinstance(raw, list):
        return [str(item) for item in raw]
    return None


def _parse_disabled(server_data: dict) -> bool:
    raw = server_data.get("disabled")
    if raw is None:
        return False
    if isinstance(raw, bool):
        return raw
    if isinstance(raw, str):
        return raw.strip().lower() in ("true", "1", "yes", "on")
    return bool(raw)


def parse_config(config_json: str) -> MCPConfig:
    # Pre-process: strip JSONC comments, then extract first JSON object
    # (some real-world files contain multiple objects or comment headers)
    preprocessed = _strip_jsonc_comments(config_json)
    preprocessed = _extract_first_json_object(preprocessed)
    try:
        data = json.loads(preprocessed)
    except json.JSONDecodeError as e:
        raise ValueError(f"Invalid JSON: {e}")

    if not isinstance(data, dict):
        raise ValueError("Config must be a JSON object")

    config_hash = hashlib.sha256(config_json.encode()).hexdigest()[:16]

    raw_servers = data.get("mcpServers", {})
    if not isinstance(raw_servers, dict):
        raise ValueError("'mcpServers' must be an object")

    servers: list[MCPServer] = []
    for name, server_data in raw_servers.items():
        if not isinstance(server_data, dict):
            continue
        servers.append(MCPServer(
            name=name,
            command=server_data.get("command"),
            args=[str(a) for a in server_data.get("args", [])],
            env={k: str(v) for k, v in server_data.get("env", {}).items()},
            headers=_parse_headers(server_data.get("headers")),
            url=server_data.get("url"),
            transport=server_data.get("transport"),
            auto_approve=_parse_auto_approve(server_data),
            disabled=_parse_disabled(server_data),
        ))

    return MCPConfig(config_hash=config_hash, servers=servers)

File: packages/cli/internal/engine/parser/parser.go
// Package parser handles MCP config parsing (claude_desktop_config.json and .cursor/mcp.json).
// It strips JSONC comments before unmarshaling, matching the Python parser.py logic exactly.
package parser

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
)

// AutoApprove holds the autoApprove / alwaysAllow field, which can be bool, string, or []string.
type AutoApprove struct {
	Bool   *bool
	Str    *string
	List   []string
}

// MCPServer represents one entry in mcpServers.
type MCPServer struct {
	Name        string
	Command     string
	Args        []string
	Env         map[string]string
	Headers     map[string]string
	URL         string
	Transport   string
	AutoApprove *AutoApprove
	Disabled    bool
}

// MCPConfig is the parsed representation of the full config file.
type MCPConfig struct {
	ConfigHash string
	Servers    []MCPServer
}

// ParseConfig parses a claude_desktop_config.json or .cursor/mcp.json string.
// Handles JSONC (comments), extracts the first JSON object, and normalises fields.
func ParseConfig(configJSON string) (*MCPConfig, error) {
	stripped := stripJSONCComments(configJSON)
	extracted := extractFirstJSONObject(stripped)

	var data map[string]json.RawMessage
	if err := json.Unmarshal([]byte(extracted), &data); err != nil {
		return nil, fmt.Errorf("invalid JSON: %w", err)
	}

	hash := fmt.Sprintf("%x", sha256.Sum256([]byte(configJSON)))
	configHash := hash[:16]

	rawServers, ok := data["mcpServers"]
	if !ok {
		return &MCPConfig{ConfigHash: configHash, Servers: nil}, nil
	}

	var serversMap map[string]json.RawMessage
	if err := json.Unmarshal(rawServers, &serversMap); err != nil {
		return nil, fmt.Errorf("'mcpServers' must be an object")
	}

	servers := make([]MCPServer, 0, len(serversMap))
	for name, raw := range serversMap {
		s, err := parseServer(name, raw)
		if err != nil {
			continue // skip malformed server entries, matching Python behaviour
		}
		servers = append(servers, s)
	}

	return &MCPConfig{ConfigHash: configHash, Servers: servers}, nil
}

func parseServer(name string, raw json.RawMessage) (MCPServer, error) {
	var d map[string]json.RawMessage
	if err := json.Unmarshal(raw, &d); err != nil {
		return MCPServer{}, err
	}

	s := MCPServer{
		Name:    name,
		Env:     map[string]string{},
		Headers: map[string]string{},
	}

	if v, ok := d["command"]; ok {
		json.Unmarshal(v, &s.Command) //nolint:errcheck
	}
	if v, ok := d["url"]; ok {
		json.Unmarshal(v, &s.URL) //nolint:errcheck
	}
	if v, ok := d["transport"]; ok {
		json.Unmarshal(v, &s.Transport) //nolint:errcheck
	}

	if v, ok := d["args"]; ok {
		var rawArgs []json.RawMessage
		if err := json.Unmarshal(v, &rawArgs); err == nil {
			for _, a := range rawArgs {
				var str string
				if json.Unmarshal(a, &str) == nil {
					s.Args = append(s.Args, str)
				}
			}
		}
	}

	if v, ok := d["env"]; ok {
		var rawEnv map[string]json.RawMessage
		if json.Unmarshal(v, &rawEnv) == nil {
			for k, val := range rawEnv {
				var str string
				if json.Unmarshal(val, &str) == nil {
					s.Env[k] = str
				}
			}
		}
	}

	if v, ok := d["headers"]; ok {
		var rawHeaders map[string]json.RawMessage
		if json.Unmarshal(v, &rawHeaders) == nil {
			for k, val := range rawHeaders {
				var str string
				if json.Unmarshal(val, &str) == nil {
					s.Headers[k] = str
				}
			}
		}
	}

	if v, ok := d["autoApprove"]; ok {
		s.AutoApprove = parseAutoApprove(v)
	} else if v, ok := d["alwaysAllow"]; ok {
		s.AutoApprove = parseAutoApprove(v)
	}

	if v, ok := d["disabled"]; ok {
		s.Disabled = parseDisabled(v)
	}

	return s, nil
}

func parseAutoApprove(raw json.RawMessage) *AutoApprove {
	var b bool
	if json.Unmarshal(raw, &b) == nil {
		return &AutoApprove{Bool: &b}
	}
	var s string
	if json.Unmarshal(raw, &s) == nil {
		return &AutoApprove{Str: &s}
	}
	var list []string
	if json.Unmarshal(raw, &list) == nil {
		return &AutoApprove{List: list}
	}
	return nil
}

func parseDisabled(raw json.RawMessage) bool {
	var b bool
	if json.Unmarshal(raw, &b) == nil {
		return b
	}
	var s string
	if json.Unmarshal(raw, &s) == nil {
		switch s {
		case "true", "1", "yes", "on":
			return true
		}
	}
	return false
}

// stripJSONCComments removes // line comments and /* */ block comments from JSONC text.
// This is a string-aware state machine that handles escaped characters inside strings
// so backslash-escaped quotes don't falsely toggle the in-string state.
func stripJSONCComments(text string) string {
	result := make([]byte, 0, len(text))
	inString := false
	i := 0
	for i < len(text) {
		ch := text[i]
		if inString {
			if ch == '\\' && i+1 < len(text) {
				// Escaped char — consume both without toggling state
				result = append(result, ch, text[i+1])
				i += 2
				continue
			}
			if ch == '"' {
				inString = false
			}
			result = append(result, ch)
		} else {
			if ch == '"' {
				inString = true
				result = append(result, ch)
			} else if ch == '/' && i+1 < len(text) {
				if text[i+1] == '/' {
					// Line comment — skip to end of line
					for i < len(text) && text[i] != '\n' {
						i++
					}
					continue
				} else if text[i+1] == '*' {
					// Block comment — skip to */
					i += 2
					for i < len(text)-1 {
						if text[i] == '*' && text[i+1] == '/' {
							i += 2
							break
						}
						i++
					}
					continue
				} else {
					result = append(result, ch)
				}
			} else {
				result = append(result, ch)
			}
		}
		i++
	}
	return string(result)
}

// extractFirstJSONObject returns the first complete {...} block from text.
// String-aware: braces inside quoted strings are not counted.
func extractFirstJSONObject(text string) string {
	depth := 0
	inString := false
	start := -1
	i := 0
	for i < len(text) {
		ch := text[i]
		if ch == '\\' && inString {
			i += 2
			continue
		}
		if ch == '"' {
			inString = !inString
		} else if !inString {
			if ch == '{' {
				if depth == 0 {
					start = i
				}
				depth++
			} else if ch == '}' {
				depth--
				if depth == 0 && start != -1 {
					return text[start : i+1]
				}
			}
		}
		i++
	}
	if start != -1 {
		return text[start:]
	}
	return text
}
